Initial sample (#1)
Azure-Samples/ms-identity-javascript-nodejs-desktop: README.md

* Configuration
* Login and logout
* Acquiring an access token
* Calling a web API

## Contents

| File/folder           | Description                                                  |
|-----------------------|--------------------------------------------------------------|
| `AppCreationScripts/` | Contains Powershell scripts for automating app registration. |
| `App/authProvider.js` | Main authentication logic resides here.                      |
| `App/main.js`         | Application main process.                                    |
| `App/fetch.js`        | Axios HTTP client for calling endpoints with a bearer token. |
| `App/renderer.js`     | Renderer processes and UI methods.                           |
| `App/constants.js`    | Example user accounts in JSON                                |
| `.env`                | Environment variables of authentication parameters.          |

## Getting Started

### Prerequisites

* [Node.js](https://nodejs.org/en/) must be installed to run this sample.
* [Visual Studio Code](https://code.visualstudio.com/download) is recommended for running and editing this sample.

### Setup

> 1. [Register a new application](https://docs.microsoft.com/azure/active-directory/develop/scenario-desktop-app-registration) in the [Azure Portal](https://portal.azure.com).

Reviewer: ```suggestion
1. Register a [desktop app that calls web APIs](https://docs.microsoft.com/azure/active-directory/develop/scenario-desktop-app-registration) in the [Azure Portal](https://portal.azure.com).
```

Reviewer: minor change to link

> 1. Clone this repository `git clone https://github.com/Azure-Samples/ms-identity-javascript-nodejs-desktop.git`

Reviewer: ```suggestion
    - Navigate to the Authentication blade in the app registration.
    - Click add a platform and choose **Mobile and Desktop application.**
    - For the **redirect URI**, choose custom redirect URI. Enter `msal://redirect` into the input box.
1. Clone this repository `git clone https://github.com/Azure-Samples/ms-identity-javascript-nodejs-desktop.git`
```

---
page_type: sample
languages:
- javascript
products:
- nodejs
- electron
- ms-graph
- azure-active-directory
description: "Demonstrates how to use MSAL Node to sign-in users and acquire access tokens for a protected resource such as Microsoft Graph in an Electron desktop application using the OAuth 2.0 authorization code flow with PKCE."
urlFragment: "ms-identity-javascript-nodejs-desktop"
---

# An Electron desktop application secured by MSAL Node (Preview) on Microsoft identity platform

> This sample demonstrates how to use [MSAL Node (Preview)](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/lib/msal-node) to sign-in a user and acquire an access token for a protected resource such as Microsoft Graph in an Electron desktop application using the [authorization code grant with PKCE](https://docs.microsoft.com/azure/active-directory/develop/v2-oauth2-auth-code-flow).

Reviewer 2: >with PKCE [](start = 315, length = 9)

with PKCE flow

> :information_source: Looking for a TypeScript implementation? See: [ElectronTestApp](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-node-samples/standalone-samples/ElectronTestApp)

## Features

This sample demonstrates the following **MSAL Node** concepts:

* Configuration
* Login and logout
* Acquiring an access token

> * Calling a web API

Reviewer 2: are these four documented somewhere? or are they links to sections in README?

| `AppCreationScripts/` | Contains Powershell scripts for automating app registration. |
| `App/authProvider.js` | Main authentication logic resides here.                      |
| `App/main.js`         | Application main process.                                    |
| `App/fetch.js`        | Axios HTTP client for calling endpoints with a bearer token. |
| `App/renderer.js`     | Renderer processes and UI methods.                           |
| `App/constants.js`    | Example user accounts in JSON                                |
| `.env`                | Environment variables of authentication parameters.          |

## Getting Started

### Prerequisites

* [Node.js](https://nodejs.org/en/) must be installed to run this sample.
* [Visual Studio Code](https://code.visualstudio.com/download) is recommended for running and editing this sample.

### Setup

1. Register a [desktop app that calls web APIs](https://docs.microsoft.com/azure/active-directory/develop/scenario-desktop-app-registration) in the [Azure Portal](https://portal.azure.com).
    - Navigate to the Authentication blade in the app registration.

>     - Click add a platform and choose **Mobile and Desktop application.**

Reviewer: ```suggestion
    - Be sure to **remove** any platforms registered here (e.g. web application, SPA, etc)
    - Click add a platform and choose **Mobile and Desktop application.**
```

    * For the **redirect URI**, enter `msal://redirect` into the input box.
1. Clone this repository `git clone https://github.com/Azure-Samples/ms-identity-javascript-nodejs-desktop.git`
1. Open the [.env](.env) file and provide the required configuration values.
    1. Replace the string `Enter_the_Application_Id_Here` with your app/client ID on Azure AD portal.
    1. Replace the string `Enter_the_Tenant_Id_Here` with your tenant ID on Azure AD portal.

>     1. Replace the string `Enter_the_Client_Secret_Here` with the client secret you created on Azure AD portal.

Reviewer 2: >Replace the string `Enter_the_Client_Secret_Here` with the client secret you created on Azure AD portal. [](start = 7, length = 104)

it does not exists

    1. Replace the string `Enter_the_Cloud_Instance_Id_Here` with `https://login.microsoftonline.com/` (see **note** below).
    1. Replace the string `Enter_the_Graph_Endpoint_Here`. with `https://graph.microsoft.com/` (see **note** below).

> > :information_source: *note*: This is for multi-tenant applications located on the global Azure cloud. For more information, see: [Use MSAL in a national cloud environment](https://docs.microsoft.com/azure/active-directory/develop/quickstart-v2-javascript-auth-code)

Reviewer 2: >global [](start = 84, length = 6)

global->Global

> > :information_source: *note*: This is for multi-tenant applications located on the global Azure cloud. For more information, see: [Use MSAL in a national cloud environment](https://docs.microsoft.com/azure/active-directory/develop/quickstart-v2-javascript-auth-code)

Reviewer 2: >global Azure cloud [](start = 84, length = 18)

link to [https://docs.microsoft.com/en-us/azure/active-directory/develop/authentication-national-cloud](https://docs.microsoft.com/en-us/azure/active-directory/develop/authentication-national-cloud)
